### PR TITLE
feat: get footer from cdn

### PR DESCRIPTION
**Pull request by moT01**
Checklist:

- [ ] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [ ] My pull request targets the `main` branch of freeCodeCamp.
- [ ] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX
ref #47174

### CODE REVIEW

File: client/i18n/config.js [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 import i18n from 'i18next';
 import { initReactI18next } from 'react-i18next';
+import fetch from 'node-fetch';
```
Comment (ShaunSHamilton) [inline client/i18n/config.js:5]: _Just upgrade us to Node 18..._ 🤫

**Comment by scissorsneedfoodtoo** [comment]
This seems to work okay locally, but I noticed a couple of things.

There are some new warnings in the terminal:

```
warn Unsafe builtin method was used, future builds will need to rebuild all
pages
warn WebpackError: Unsafe builtin usage https.request:
    at @freecodecamp/node_modules/node-fetch/lib/index.mjs:1462:1
    at @freecodecamp/node_modules/node-fetch/lib/index.mjs:1431:1
...
```

And I'm pretty sure `npm run build` completes even when the URL to the trending YAML file is wrong. Throwing in `process.exit(1)` in `catch` just after logging the error seems to kill the build, but it's through Webpack saying that `process.exit` isn't a function.

Not sure if those warnings are something we can safely ignore or not, or if there's a better way to stop the build if the CDN is down, the YAML file for the current locale doesn't exist, etc.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
It might be better to expand the `create:config` script so that it downloads the yaml and converts it to JSON.  That way, yes, there's a special case, but this file can just handle that json file exactly like the rest.

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams just to clarify, we add another script `download:trending` to `create:config` which to downloads the trending articles converts them to JSON and save them in ./locales/${clientLocale}/trending.json ?

**Comment by ahmadabdolsaheb** [comment]
Not sure what would be the best way to test this, should I async pull all trending and see if they have all keys?

**Comment by ojeytonwilliams** [comment]
> Not sure what would be the best way to test this, should I async pull all trending and see if they have all keys?

I recommend a couple of things.

Firstly, the pipeline should collapse if there's an error. We can implement a retry strategy if that happens a lot.

Secondly, yeah, validating the shape of the data immediately after download makes sense.  Something like the `joi` tests in `challengeSchema`.  Also, if the number of keys is constant, we can encode that in the schema.

**Comment by scissorsneedfoodtoo** [comment]
Hey all, I was able to circle back around to this and write up some schema validation for the trending data. I pushed it to a branch on my fork for now in case we wanted to go in a different direction: https://github.com/freeCodeCamp/freeCodeCamp/compare/main...scissorsneedfoodtoo:freeCodeCamp:feat/get-footer-from-cdn

It should be possible to download the trending YAML files for each of the active locales and stick with the current tests, but we may run into issues where a file is missing, or the schema for one locale is messed up, and all the builds would get blocked until that's fixed.

I figure it might be better to remove the existing `translation.json` schema and file tests in favor of `Joi` schema validation, though only for whatever `clientLocale` is set to.

Also, I found that it's possible to validate the trending data against a schema with dynamic keys like this:

```js
const schema = Joi.object().min(60).max(60).pattern(
  /article\d+(title|link)/, [Joi.string().required(), Joi.string().uri({ scheme: 'https' }).required().required()]
);
```

The only issue is that, since the trending data is a flat object, it's possible to have a mixed up title and link and still pass validation. I don't think it's a huge issue, though, so we could go with the fancier schema above rather than a large object that checks each article and title individually.

If all this sounds okay, just let me know and I can push back to this branch.

**Review by axljoker03 — COMMENTED** [review]
thank

**Comment by ojeytonwilliams** [comment]
> The only issue is that, since the trending data is a flat object, it's possible to have a mixed up title and link and still pass validation. I don't think it's a huge issue, though, so we could go with the fancier schema above rather than a large object that checks each article and title individually.

That sounds reasonable to me.  That said, joi supports `.not(` so you could validate that titles don't start with `https://`. Unless, of course, some titles actually do start with that!

Either way, the approach you've described seems reasonable.

**Comment by scissorsneedfoodtoo** [comment]
@moT01, hope you doin't mind, but I went ahead and pushed some updates to your branch.

@ojeytonwilliams, this should be ready for review now. I think I was able to find a way to test article tittles and links separately so only article links should match the https URI scheme.

About the titles, none on the English publication include `https://`, though I suppose it could happen! It seemed better to have a looser test for the title strings, at least for now.

File: client/i18n/schema/trendingSchema.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import Joi from 'joi';
+
+const schema = Joi.object()
+  .min(60)
+  .max(60)
+  .pattern(/article\d+title/, [Joi.string().required()])
```
Comment (ahmadabdolsaheb) [inline client/i18n/schema/trendingSchema.js:6]: This is solid and we could possibly make it more bullet proof if we added an iterator for the keys to add the numbers in the keys or hardcode the keys such as the following:
```
Joi.object().keys({
    article0title: Joi.string().required(),
    article0link: Joi.string().uri({ scheme: 'https' }).required(),
...
})
```

Comment (scissorsneedfoodtoo) [reply client/i18n/schema/trendingSchema.js:6]: Hey @ahmadabdolsaheb, sure, we could do that. I don't see the harm in having a really plain, easy to read schema for something like this. I seriously doubt we'll be updating the schema anytime soon.

I'll go in and revert to hardcoded keys / values for the schema validation.

**Review by ahmaxed — APPROVED** [review]

**Review by raisedadead — APPROVED** [review]

**Comment by scissorsneedfoodtoo** [comment]
@ojeytonwilliams, could you take a look at this when you get a chance? This should be ready for review.

I tried not to rebase after the last couple of reviews, though I did mess up some scripts that @moT01 and @ahmadabdolsaheb had in `package.json`. Everything should be sorted now.

And I can work on squashing these commits if that makes things a bit easier.

File: tools/scripts/build/download-trending.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { writeFileSync } from 'fs';
+import fetch from 'node-fetch';
+import yaml from 'js-yaml';
+import { trendingSchemaValidator } from '../../../client/i18n/schema/trendingSchema';
+
+import envData from '../../../config/env.json';
+const { clientLocale } = envData;
+
+const cdnUrlCreator = (lang: string) =>
```
Comment (ojeytonwilliams) [inline tools/scripts/build/download-trending.ts:9]:
```suggestion
const createCdnUrl = (lang: string) =>
```
nit it's a function not an object :man_shrugging:

File: tools/scripts/build/download-trending.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  `https://cdn.freecodecamp.org/universal/trending/${lang}.yaml`;
+
+(async () => {
```
Comment (ojeytonwilliams) [inline tools/scripts/build/download-trending.ts:12]:
```suggestion
const download async () => {
```
We don't need to use process.exit if we go this way

File: tools/scripts/build/download-trending.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    Error: The trending JSON is invalid.
+    ----------------------------------------------------
+    Unable to validate the ${clientLocale} trending JSON schema: ${validationError.message}
+    `
+    );
+  }
+})().catch(err => {
+  console.log(err);
+  process.exit(1);
+});
```
Comment (ojeytonwilliams) [inline tools/scripts/build/download-trending.ts:48]:
```suggestion
};

void download();
```

File: tools/scripts/build/download-trending.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import { trendingSchemaValidator } from '../../../client/i18n/schema/trendingSchema';
+
+import envData from '../../../config/env.json';
+const { clientLocale } = envData;
+
+const cdnUrlCreator = (lang: string) =>
+  `https://cdn.freecodecamp.org/universal/trending/${lang}.yaml`;
+
+(async () => {
+  const url = cdnUrlCreator(clientLocale);
```
Comment (ojeytonwilliams) [inline tools/scripts/build/download-trending.ts:13]:
```suggestion
  const url = createCdnUrl(clientLocale);
```

**Review by ojeytonwilliams — APPROVED** [review]
I only found tiny style issues, so it can go in as is.  Let me know what you think @scissorsneedfoodtoo

**Comment by scissorsneedfoodtoo** [comment]
Awesome, thanks for taking a look at this @ojeytonwilliams, and for all of those suggestions. All of those LGTM.

I'll go ahead and commit them now 👍

**Review by ojeytonwilliams — APPROVED** [review]
Nice, I think this is good to go :rocket:

**Pull request merged**